core/src/components/router/router.tsx

```diff
@@ -61,7 +61,17 @@ export class Router implements ComponentInterface {
     await waitUntilNavNode();
     console.debug('[ion-router] found nav');
 
-    await this.onRoutesChanged();
+    const canProceed = await this.runGuards(this.getPath());
+    if (canProceed !== true) {
+      if (typeof canProceed === 'object') {
+        const path = parsePath(canProceed.redirect);
+        const queryString = canProceed.redirect.split('?')[1];
```

Comment:
What about including everything after the first `?` instead ?

It should be equivalent in most cases but it would avoid users shooting themselves if they include a non-escape `?` in the query string of in the fragment i.e `site.com?value=ab?cd#de?f`.

Reply:
Fixed!